node-bootstrapper/main.go

```diff
@@ -0,0 +1,396 @@
+			return fmt.Errorf("write file %s: %w", path, err)
+		}
+	}
+	return nil
+}
+
+type File struct {
+	Content string
+	Mode    os.FileMode
+}
+
+func customData(config *datamodel.NodeBootstrappingConfiguration) (map[string]File, error) {
+	contentDockerDaemon, err := contentDockerDaemonJSON(config)
+	if err != nil {
+		return nil, fmt.Errorf("content docker daemon json: %w", err)
+	}
+
+	files := map[string]File{
+		"/etc/kubernetes/certs/ca.crt": {
+			Content: config.ContainerService.Properties.CertificateProfile.CaCertificate,
+			Mode:    ReadOnlyUser,
+		},
+		"/etc/systemd/system/docker.service.d/exec_start.conf": {
+			Content: contentDockerExecStart(config),
+			Mode:    ReadOnlyWorld,
+		},
+		"/etc/docker/daemon.json": {
+			Content: contentDockerDaemon,
+			Mode:    ReadOnlyWorld,
+		},
+		"/etc/default/kubelet": {
+			Content: contentKubelet(config),
+			Mode:    ReadOnlyWorld,
+		},
+	}
+
+	if config.ContainerService.Properties.SecurityProfile.GetPrivateEgressContainerRegistryServer() != "" {
+		files["/etc/containerd/certs.d/mcr.microsoft.com/hosts.toml"] = File{
+			Content: containerDMCRHosts(config),
+			Mode:    ReadOnlyWorld,
+		}
+	}
+
+	if config.EnableSecureTLSBootstrapping || agent.IsTLSBootstrappingEnabledWithHardCodedToken(config.KubeletClientTLSBootstrapToken) {
+		if err := useBootstrappingKubeConfig(config, files); err != nil {
+			return nil, err
+		}
+	} else {
+		useHardCodedKubeconfig(config, files)
+	}
+
+	for path, file := range files {
+		file.Content = strings.TrimLeft(file.Content, "\n")
+		files[path] = file
+	}
+
+	return files, nil
+}
+
+func useHardCodedKubeconfig(config *datamodel.NodeBootstrappingConfiguration, files map[string]File) {
+	files["/var/lib/kubelet/kubeconfig"] = File{
+		Content: contentKubeconfig(config),
+		Mode:    ReadOnlyWorld,
+	}
+}
+
+func useBootstrappingKubeConfig(config *datamodel.NodeBootstrappingConfiguration, files map[string]File) error {
+	bootstrapKubeconfig, err := contentBootstrapKubeconfig(config)
+	if err != nil {
+		return fmt.Errorf("content bootstrap kubeconfig: %w", err)
+	}
+	files["/var/lib/kubelet/bootstrap-kubeconfig"] = File{
+		Content: bootstrapKubeconfig,
+		Mode:    ReadOnlyWorld,
+	}
+	return nil
+}
+
+func contentKubeconfig(config *datamodel.NodeBootstrappingConfiguration) string {
+	users := `- name: client
+  user:
+    client-certificate: /etc/kubernetes/certs/client.crt
+    client-key: /etc/kubernetes/certs/client.key`
+
+	return fmt.Sprintf(`
+apiVersion: v1
+kind: Config
+clusters:
+- name: localcluster
+  cluster:
+    certificate-authority: /etc/kubernetes/certs/ca.crt
+    server: https://%s:443
+users:
+%s
+contexts:
+- context:
+    cluster: localcluster
+    user: client
+  name: localclustercontext
+current-context: localclustercontext
+`, agent.GetKubernetesEndpoint(config.ContainerService), users)
+}
+
+func contentBootstrapKubeconfig(config *datamodel.NodeBootstrappingConfiguration) (string, error) {
+	data := map[string]any{
```

Comment:
for example, the generated string is really hard to grasp from the code. Having an external template file would help quickly identify what the final file content looks like (formatting too) and which parameters are coming from outside.

Reply:
I hate YAML-generation in templates. Especially when IF conditions are involved.
It gives me HELM vibes with hacks around proper indentation. Tabs vs spaces etc.

In this function I intentionally avoided using templates due to its complexity.